Modify rule S5334: Change text to the education framework format [Javascript][APPSEC-235] (#1387)
SonarSource/rspec: rules/S5334/javascript/how-to-fix-it/mongo.adoc

}
run().catch(console.dir);
----

==== Compliant solution

[source,javascript,diff-id=1,diff-type=compliant]
----
const client = new MongoClient(mongoURI);
async function run() {
    const allowed = ["name", "price", "stock"]

    let propId = req.query.propId
    const database = client.db('example');
    const products = database.collection('product');
    const query = { $where: `isString(this.${allowed[propId]})`};
    const product = await products.findOne(query); // Noncompliant
}
run().catch(console.dir);
----

=== How does this work?

MongoDB provides JavaScript code-aware search operators to allow developers to
perform advanced queries. In most cases, the same result can be achieved by
combining classical, safer, operators, or by performing additional data
filtering on the application side.

Note that MongoDB servers implement a sandboxed JavaScript engine to run the
operator-related code. Without other vulnerabilities in the database engine,
the execution of arbitrary code is thus not possible through the exploitation
of evaluation operators. However, a successful injection could still lead to
the compromise of the MongoDB server hosted data. Moreover, the discovery by

> an attacker of a sandbox escaping attack payload should also be considered.

Reviewer: Not sure all users are familiar with the notion of payload. What about, 
```suggestion
an attacker of a way to escape the sandbox should also be considered.
```
